src/main/java/org/opensearch/neuralsearch/processor/DocumentChunkingProcessor.java

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.processor;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import lombok.extern.log4j.Log4j2;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.env.Environment;
+import org.opensearch.index.IndexService;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.index.analysis.AnalysisRegistry;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.indices.IndicesService;
+import org.opensearch.index.IndexSettings;
+import org.opensearch.ingest.AbstractProcessor;
+import org.opensearch.ingest.IngestDocument;
+import org.opensearch.ingest.Processor;
+import org.opensearch.neuralsearch.processor.chunker.ChunkerFactory;
+import org.opensearch.neuralsearch.processor.chunker.FieldChunker;
+import org.opensearch.index.mapper.IndexFieldMapper;
+import org.opensearch.neuralsearch.processor.chunker.FixedTokenLengthChunker;
+import static org.opensearch.ingest.ConfigurationUtils.readMap;
+import static org.opensearch.neuralsearch.processor.chunker.ChunkerFactory.DELIMITER_ALGORITHM;
+import static org.opensearch.neuralsearch.processor.chunker.ChunkerFactory.FIXED_LENGTH_ALGORITHM;
+
+/**
+ * This processor is used for chunking user input data text embedding processing, algorithm can be used to indicate chunking algorithm and parameters,
+ * and field_map can be used to indicate which fields needs chunking and the corresponding keys for the chunking results.
+ */
+@Log4j2
+public final class DocumentChunkingProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "chunking";
+
+    public static final String FIELD_MAP_FIELD = "field_map";
+
+    public static final String ALGORITHM_FIELD = "algorithm";
+
+    public static String MAX_CHUNK_LIMIT_FIELD = "max_chunk_limit";
+
+    private static final int DEFAULT_MAX_CHUNK_LIMIT = -1;
+
+    private int current_chunk_count = 0;
```

Comment:
Processor itself should be a singleton and stateless, defining this to class level could have multithreading issue, did you tested on this?

Reply:
I see. I will move currentChunkCount into the `execute` method.

Reply:
How about creating a class to for `chunkMapType(sourceAndMetadataMap, fieldMap)`?

Reply:
Didn't get your point? Creating a class for a method?

Reply:
After second thought, I implemented this logic with an inner class `ChunkCountWrapper`. Please review this part.

Reply:
Creating a class for `chunkMapType` is also a valid option. The class will maintain a member variable named `chunkCount`. But For simplicity, can we accept the wrapper solution?